File: web/blocks/external-storage.dart
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

part of NetTango;

class ExternalStorage {
  static js.JsArray getKeys(js.JsObject item) =>
    js.context["Object"].callMethod("keys", [item]);

  final List<String> propertyNames;
  js.JsObject storage;

  void set(js.JsObject storage) { this.storage = storage; }

  ExternalStorage(List<String> this.propertyNames);

  restore(js.JsObject export) {
    if (this.storage == null) {
      return;
    }
    for (String propertyName in ExternalStorage.getKeys(this.storage)) {
      if (!this.propertyNames.contains(propertyName)) {
        if (export.hasProperty(propertyName)) {
          throw new Exception("Found existing property when restoring external data for export: ${propertyName}");
        }
        export[propertyName] = storage[propertyName];
      }
    }
  }
}

File: src/utils/external-storage.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

class ExternalStorage {

  propertyNames: string[]
  storage: any | null = null

  set(storage: any) {
    if (storage === null || storage === undefined) {
      throw new Error("Cannot set storage to null or undefined.")
    }
    if (typeof(storage) !== "object") {
      throw new Error("Storage must be an object.")
    }
    this.storage = storage
  }

  constructor(propertyNames: string[]) {
    this.propertyNames = propertyNames
  }

  restore(exports: any) {
    if (this.storage === null) {
      // TODO: Hmm, why would we restore without setting the storage?  Maybe handle this better.
      // throw new Error("Cannot restore if the storage is unset.")
      return null
    }
    for (var propertyName of Object.keys(this.storage)) {
      if (!this.propertyNames.includes(propertyName)) {
        if (exports.hasOwnProperty(propertyName)) {
          throw new Error(`Found existing property when restoring external data for export: ${propertyName}`)
        }
        exports[propertyName] = this.storage[propertyName]
      }
    }
  }
}

export { ExternalStorage }
